index.js

```diff
@@ -16,114 +16,114 @@ function helmetPlugin (fastify, options, next) {
     fastify.decorateReply('helmet', null)
   }
 
-  // We will add the onRequest helmet middleware functions through the onRoute hook if needed
+  // We initialize the `cspNonce` reply decorator only if it does not already exists
+  if (!fastify.hasReplyDecorator('cspNonce')) {
+    fastify.decorateReply('cspNonce', null)
+  }
+
   fastify.addHook('onRoute', (routeOptions) => {
     if (typeof routeOptions.helmet !== 'undefined') {
       if (typeof routeOptions.helmet === 'object') {
-        const { enableCSPNonces: enableRouteCSPNonces, ...helmetRouteConfiguration } = routeOptions.helmet
-
-        // If route helmet options are set they overwrite the global helmet configuration
-        const mergedHelmetConfiguration = Object.assign({}, globalConfiguration, helmetRouteConfiguration)
-
-        buildRouteHooks(mergedHelmetConfiguration, routeOptions)
-
-        if (enableRouteCSPNonces) {
-          routeOptions.onRequest.push(buildCSPNonce(fastify, mergedHelmetConfiguration))
-        }
+        routeOptions.config = Object.assign(routeOptions.config || Object.create(null), { helmet: routeOptions.helmet })
       } else if (routeOptions.helmet === false) {
-        // don't apply any helmet settings but decorate the reply with a fallback to the
-        // global helmet options
-        buildRouteHooks(globalConfiguration, routeOptions, true)
+        routeOptions.config = Object.assign(routeOptions.config || Object.create(null), { helmet: { skipRoute: true } })
       } else {
         throw new Error('Unknown value for route helmet configuration')
       }
-    } else if (isGlobal) {
-      // if the plugin is set globally (meaning that all the routes will be decorated)
-      // As the endpoint, does not have a custom helmet configuration, use the global one.
-      buildRouteHooks(globalConfiguration, routeOptions)
+    }
+  })
 
-      if (enableCSPNonces) {
-        routeOptions.onRequest.push(buildCSPNonce(fastify, globalConfiguration))
-      }
+  fastify.addHook('onRequest', async (request, reply) => {
```

Comment:
I think you can combine the two `onRequest` hook together.

Reply:
The first iteration I wrote of this implementation was doing all the work in one hook but strangely it had a huge impact on performance. That's why I split it in two separate hooks.

For reference: https://github.com/darkgl0w/fastify-helmet/tree/single-hook-version if you want to take a look ^^

Reply:
I will try to run a few comparative `clinic.js` diagnosis to understand from where this performance drop is coming :d